pyacq/viewers/qtimefreq.py

```diff
@@ -0,0 +1,739 @@
+    f_stop: float
+        Last frequency in Hz
+    deltafreq:
+        Frequency interval in Hznew_head
+    sampling_rate:
+        Sampling rate in Hz
+    
+    Returns:
+    ----------
+        wf : Fourier transform of the wavelet coefficients (after weighting)
+              axis 0 is time,  axis 1 is freq
+    """
+    # compute final map scales
+    scales = f0/np.arange(f_start,f_stop,deltafreq)*sampling_rate
+    
+    # compute wavelet coeffs at all scales
+    xi=np.arange(-len_wavelet/2.,len_wavelet/2.)
+    xsd = xi[:,np.newaxis] / scales
+    wavelet_coefs=np.exp(complex(1j)*2.*np.pi*f0*xsd)*np.exp(-np.power(xsd,2)/2.)
+
+    weighting_function = lambda x: x**(-(1.0+normalisation))
+    wavelet_coefs = wavelet_coefs*weighting_function(scales[np.newaxis,:])
+
+    # Transform the wavelet into the Fourier domain
+    wf=scipy.fftpack.fft(wavelet_coefs,axis=0)
+    wf=wf.conj()
+    
+    return wf
+
+
+
+class ComputeThread(QtCore.QThread):
+    """
+    Thread used internal by TimeFreqWorker.
+    """
+    def __init__(self, in_stream, out_stream, channel, local, parent=None):
+        QtCore.QThread.__init__(self, parent)
+        self.in_stream = weakref.ref(in_stream)
+        self.out_stream = weakref.ref(out_stream)
+        self.channel = channel
+        self.local = local
+        
+        self.worker_params = None
+
+    def run(self):
+        if self.worker_params is None: 
+            return
+        head = self.head
+        
+        downsampling_factor = self.worker_params['downsampling_factor']
+        sig_chunk_size = self.worker_params['sig_chunk_size']
+        filter_a = self.worker_params['filter_a']
+        filter_b = self.worker_params['filter_b']
+        wavelet_fourrier = self.worker_params['wavelet_fourrier']
+        plot_length = self.worker_params['plot_length']
+        
+        #~ t1 = time.time()
+        
+        if downsampling_factor>1:
+            head = head - head%downsampling_factor
+        full_arr = self.in_stream().get_array_slice(head, sig_chunk_size)[self.channel, :]
+        
+        #~ t2 = time.time()
+        
+        if downsampling_factor>1:
+            small_arr = scipy.signal.filtfilt(filter_b, filter_a, full_arr)
+            small_arr =small_arr[::downsampling_factor].copy()# to ensure continuity
+        else:
+            small_arr = full_arr
+        
+        small_arr_f=scipy.fftpack.fft(small_arr)
+        if small_arr_f.shape[0] != wavelet_fourrier.shape[0]: return
+        wt_tmp=scipy.fftpack.ifft(small_arr_f[:,np.newaxis]*wavelet_fourrier,axis=0)
+        wt = scipy.fftpack.fftshift(wt_tmp,axes=[0])
+        wt = np.abs(wt).astype('float32')
+        wt = wt[-plot_length:]
+        
+        #~ self.last_wt_map = wt
+        self.out_stream().send(head, wt)
+        #~ t3 = time.time()
+        
+        #print('compute', self.channel,  t2-t1, t3-t2, t3-t1, QtCore.QThread.currentThreadId())
+
+
+
+class TimeFreqWorker(Node, QtCore.QObject):
```

Comment:
At the moment there is not special category of Node except the dirname they are.

Yes this TimeFreqWorker could be somewhere else. We will see when re implementing with vispy.